include/units/core.h

```diff
@@ -165,10 +165,91 @@ namespace units
 #define UNIT_ADD_UNIT_DEFINITION(namespaceName, nameSingular) \
 	inline namespace namespaceName \
 	{ \
-		/** @name Unit Containers */ /** @{ */ template<class Underlying> \
-		using nameSingular##_t = unit<nameSingular, Underlying>; /** @} */ \
+		/** @name Unit Containers */ /** @{ */ UNIT_ADD_SCALED_UNIT_DEFINITION( \
+			nameSingular##_t, ::units::linear_scale, nameSingular) /** @} */ \
 	}
 
+/**
+ * @def			UNIT_ADD_SCALED_UNIT_DEFINITION(unitName, scale, definition)
+ * @brief		Macro for generating the boiler-plate code for the scaled unit template definition.
+ * @details		The macro generates the definition of the scaled unit templates as a strong type template alias,
+ *				e.g. `meters`
+ * @param		unitName unit name, e.g. 'meters'
+ * @param		scale the non linear scale template argument of the unit's base
+ * @param		definition - the variadic parameter is used for the definition of the unit
+ *				(e.g. `conversion_factor<std::ratio<1>, units::dimension::length>`)
+ * @note		a variadic template is used for the definition to allow templates with
+ *				commas to be easily expanded. All the variadic 'arguments' should together
+ *				comprise the unit definition.
+ */
+#define UNIT_ADD_SCALED_UNIT_DEFINITION(unitName, scale, /*definition*/...) \
+	template<class Underlying> \
+	class unitName : public ::units::unit<__VA_ARGS__, Underlying, scale> \
+	{ \
+		using _base = ::units::unit<__VA_ARGS__, Underlying, scale>; \
+\
+	public: \
+		constexpr unitName()                = default; \
+		constexpr unitName(const unitName&) = default; \
+		constexpr unitName(unitName&&)      = default; \
+		constexpr unitName& operator=(const unitName&) = default; \
+		constexpr unitName& operator=(unitName&&) = default; \
+\
+		template<class T, class... Args, \
+			::std::enable_if_t<::std::is_constructible_v<_base, T, Args...> && \
+					(sizeof...(Args) != 0 || !::std::is_convertible_v<T, _base>), \
+				int> = 0> \
+		explicit constexpr unitName(T&& val, Args&&... args) noexcept \
+		  : _base(::std::forward<T>(val), ::std::forward<Args>(args)...) \
+		{ \
+		} \
+\
+		template<class T, \
+			::std::enable_if_t<::std::is_constructible_v<_base, T>&& ::std::is_convertible_v<T, _base>, int> = 0> \
+		constexpr unitName(T&& val) noexcept : _base(::std::forward<T>(val)) \
+		{ \
+		} \
+\
+		template<class T, ::std::enable_if_t<::std::is_assignable_v<_base&, T>, int> = 0> \
+		constexpr unitName& operator=(T&& val) noexcept \
+		{ \
+			static_cast<_base&>(*this) = ::std::forward<T>(val); \
+			return *this; \
+		} \
+\
+		template<class Unit = unitName> \
+		constexpr const char* name() const noexcept \
+		{ \
+			return ::units::unit_name_v<Unit>; \
+		} \
+\
+		template<class Unit = unitName> \
+		constexpr const char* abbreviation() const noexcept \
+		{ \
+			return ::units::unit_abbreviation_v<Unit>; \
+		} \
+	}; \
+\

```

Comment:
```cpp
\
	unitName()->unitName<UNIT_LIB_DEFAULT_TYPE>; \
```

I want a template deduction guide for the default constructor that uses the libraries default type. This will provide backwards compatibility to 2.3 code (yes I'm ignoring the fact that we're going to change all the names too), and also it keeps things simple for the template-averse or in situations where the underlying type isn't very important.

Reply:
Very well. I'll first rebase and then add a commit with that. IIRC from previous tries:
- That'll require braces, like in `meter_t{}`, as `meter_t` is ill-formed.
- Adding that to the `Underlying` class template parameter made all units have `double` underlying type, so the solution above would still require using `meter_t<double>` for function arguments and such.